drivers/xen/regions.c

```diff
@@ -0,0 +1,191 @@
+		     (DT_REG_ADDR_BY_IDX(XEN_HYP_NODE, idx) +		\
+		      DT_REG_SIZE_BY_IDX(XEN_HYP_NODE, idx)) >		\
+		     CONFIG_KERNEL_VM_BASE + CONFIG_KERNEL_VM_SIZE),	\
+		     "Extended region should not overlap virtual ram");	\
+	SYS_MEM_BLOCKS_DEFINE_STATIC_WITH_EXT_BUF(			\
+		region_##idx, XEN_PAGE_SIZE,				\
+		XEN_PFN_DOWN(DT_REG_SIZE_BY_IDX(XEN_HYP_NODE, idx)),	\
+		(uint8_t *)DT_REG_ADDR_BY_IDX(XEN_HYP_NODE, idx));
+
+DT_FOREACH_REG(XEN_HYP_NODE, EXT_BLOCK);
+
+#define EXT_REG_POINTER(node_id, idx)					\
+	{								\
+		.block = &region_##idx,					\
+		.phys_addr = DT_REG_ADDR_BY_IDX(XEN_HYP_NODE, idx),	\
+		.size = DT_REG_SIZE_BY_IDX(XEN_HYP_NODE, idx),		\
+	},
+
+/*
+ * TODO: For now we allocate buffer structures for DT_NUM_REGS, when we
+ * need a range 1...DT_FOREACH_REG. Although region 0 is ignored, it still
+ * takes space for sys_block structures allocation. This requires optimization
+ */
+static struct extended_regions ext_blocks[DT_NUM_REGS(XEN_HYP_NODE)] = {
+	DT_FOREACH_REG(XEN_HYP_NODE, EXT_REG_POINTER)
+};
+
+void *xen_region_get_pages(size_t nr_pages)
+{
+	int ret, i;
+	void *ptr;
+
+	/*
+	 * Locking ranges iteration, although each call sys_mem_blocks_alloc_contiguous
+	 * has it's own syncronisation, to avoid the situation of 2 parralel calls of
+	 * get_page and put_page on full buffers when put_page clears space in block 1
+	 * when get_page has already checked block 1. This will lead us to -ENOMEM while
+	 * block 1 space was freed already
+	 */
+	k_mutex_lock(&ext_regions_lock, K_FOREVER);
+
+	for (i = EXTENDED_REGIONS_IDX; i < DT_NUM_REGS(XEN_HYP_NODE); i++) {
+		ret = sys_mem_blocks_alloc_contiguous(ext_blocks[i].block, nr_pages, &ptr);
+		if (ret == 0) {
+			break;
+		} else if (ret && ret != -ENOMEM) {
+			LOG_ERR("Unable to allocate pages, ret = %d\n", ret);
+			ptr = NULL;
+		}
+	}
+
+	k_mutex_unlock(&ext_regions_lock);
+
+	return ptr;
+}
+
+static int addr_is_in_buffer(uint8_t *ptr, sys_mem_blocks_t *block)
+{
+	/* Compare ptr to buffer address to determine if it was allocated in this block */
+	return (ptr >= block->buffer) &&
+		(ptr < block->buffer + (block->num_blocks << block->blk_sz_shift));
+}
+
+void xen_region_put_pages(void *ptr, size_t nr_pages)
+{
+	int ret, i;
+	sys_mem_blocks_t *block = NULL;
+
+	k_mutex_lock(&ext_regions_lock, K_FOREVER);
+
+	for (i = EXTENDED_REGIONS_IDX; i < DT_NUM_REGS(XEN_HYP_NODE); i++) {
+		if (addr_is_in_buffer(ptr, ext_blocks[i].block)) {
+			block = ext_blocks[i].block;
+		}
+	}
+
+	if (!block) {
+		LOG_ERR("Unable to find buffer, matching to addr %p\n", ptr);
+		goto unlock;
+	}
+
+	ret = sys_mem_blocks_free_contiguous(block, ptr, nr_pages);
+	if (ret) {
+		LOG_ERR("Unable to free pages, ret = %d\n", ret);
+		goto unlock;
+	}
+
+unlock:
+	k_mutex_unlock(&ext_regions_lock);
+}
+
+static int xen_regions_init(const struct device *d)
+{
+	int i;
+
+	ARG_UNUSED(d);
+
+	LOG_DBG("Initializing Xen Extended regions driver");
+
+	k_mutex_lock(&ext_regions_lock, K_FOREVER);
```

Comment:
I think this lock is redundant here, at this stage no one is able to access these regions.